swarm/src/connection/handler_wrapper.rs

```diff
@@ -243,6 +247,11 @@ where
     ) {
         match endpoint {
             SubstreamEndpoint::Listener => {
+                if self.negotiating_in.len() == MAX_NUM_NEGOTIATING_IN {
```

Comment:
Instead of doing this check here, we could create a wrapper around `FuturesUnordered` that enforces a limit on the number of concurrent futures. This would foster reuse and actually enforce this limit. At the moment, there is a risk that this check is circumvented in case the `FuturesUnordered` is modified in a different place without this check.

Reply:
:+1: for the reusable abstraction, though I would much rather like to remove the entire upgrade mechanism for substreams in the future.